### PR TITLE
Always emit an error for a query cycle

### PR DESCRIPTION
**Pull request by Zoxc**
r? @michaelwoerister 

cc @nikomatsakis @wesleywiser

### CODE REVIEW

File: src/librustc/ty/steal.rs [hunk 1]
```diff
@@ -43,4 +43,8 @@ impl<T> Steal<T> {
         let value = value_ref.take();
         value.expect("attempt to read from stolen value")
     }
+
+    pub fn stolen(&self) -> bool {
+        self.value.borrow().is_some()
```
[the remark this answers is not in the document]
Comment (Zoxc) [reply src/librustc/ty/steal.rs]: Fixed

**Comment by Dylan-DPC** [comment]
ping from triage @Zoxc you have conflicts and failing tests to resolve

**Comment by michaelwoerister** [comment]
Sorry, I'm behind on reviewing this...

File: src/librustc/ty/steal.rs [hunk 2]
```diff
@@ -43,4 +43,8 @@ impl<T> Steal<T> {
         let value = value_ref.take();
         value.expect("attempt to read from stolen value")
     }
+
+    pub fn stolen(&self) -> bool {
```
Comment (michaelwoerister) [inline src/librustc/ty/steal.rs]: Is this used anywhere? I don't see it anywhere else in the diff.

**Review by michaelwoerister — COMMENTED** [review]
The PR looks good. @rust-lang/compiler who is looking into MIR inlining at the moment? There are some changes in here that might affect it (but I assume that the current solution is temporary anyway?).

**Comment by wesleywiser** [comment]
@michaelwoerister I've played around with it a bit in the past. FYI, there's some more context for this PR here https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/always.20error.20on.20query.20cycles

File: src/librustc/ty/steal.rs (hunk 2 above)
Comment (Zoxc) [reply src/librustc/ty/steal.rs]: It was used in a previous version. I'll remove it.

**Comment by michaelwoerister** [comment]
OK, r=me with the nits addressed.

File: src/librustc_mir/transform/inline.rs [hunk 3]
```diff
@@ -98,22 +98,34 @@ impl<'a, 'tcx> Inliner<'a, 'tcx> {
...
-                            param_env,
-                            callee_mir,
-                        )
+                let self_node_id = self.tcx.hir().as_local_node_id(self.source.def_id()).unwrap();
+                let callee_node_id = self.tcx.hir().as_local_node_id(callsite.callee);
+
+                let callee_mir = if let Some(callee_node_id) = callee_node_id {
+                    // Avoid a cycle here by only using `optimized_mir` only if we have
+                    // a lower node id than the callee. This ensures that the callee will
+                    // not inline us. This trick only works without incremental compilation.
```
[the remark this answers is not in the document]
Comment (Zoxc) [reply src/librustc_mir/transform/inline.rs]: We do not want optimizations to vary though.

**Pull request merged**